File: unittest/WaveTable_test.cc

#include <boost/test/auto_unit_test.hpp>
#include "boost/filesystem/operations.hpp"
#include "boost/filesystem/fstream.hpp"   
#include <iostream>                        
#include <fstream>
#include "h5filerecorder.h"
#include "wavetable.h"
using namespace soma; 
using  namespace boost;       
using namespace boost::filesystem; 
using namespace std; 

BOOST_AUTO_TEST_SUITE(WaveTableTest); 

BOOST_AUTO_TEST_CASE(WaveTable_create)
{

  // create a temp file
  H5::H5File::H5File * h5file 
    = new H5::H5File::H5File("WaveTable_create.h5", H5F_ACC_TRUNC); 

  H5::Group grp = h5file->createGroup("testGroup");
  
  int SRC = 17; 
  WaveTable tst(SRC, "eeg1",  grp); 
  h5file->flush(H5F_SCOPE_GLOBAL); 

  h5file->close(); 
  delete h5file; 

  int retval = system("python WaveTable_test.py create");
  BOOST_CHECK_EQUAL(retval , 0); 

}

BOOST_AUTO_TEST_CASE(WaveTable_append)
{
  // create a temp file
  H5::H5File::H5File * h5file 
    = new H5::H5File::H5File("WaveTable_append.h5", H5F_ACC_TRUNC); 

  H5::Group grp = h5file->createGroup("testGroup");
  
  int SRC = 23; 
  WaveTable*  tst = new WaveTable(23, "eeg2", grp); 
  int N = 1000; 
  
  uint64_t offsetval; 
  offsetval = 0x01234567; 
  offsetval = offsetval << 16; 
  offsetval |= 0x89AB; 
  for (uint64_t i = 0; i < N; i++)
    {
      Wave_t w; 
      w.src = SRC; 
      w.time = i * offsetval;
      w.samprate = 0x789A; 
      w.selchan = 0x1234; 
      w.filterid = 0xAABB; 
      
      for (int j = 0; j < WAVEBUF_LEN; j++) {
	w.wave[j] = i * 0xABCD + j*0x17; 
      }
      
      pDataPacket_t rdp = rawFromWave(w); 
      tst->append(rdp); 
      rdp.reset(); 


    }
  delete tst; 
  h5file->flush(H5F_SCOPE_GLOBAL); 

  h5file->close(); 
  h5file->flush(H5F_SCOPE_GLOBAL); 

  delete h5file; 


  int retval = system("python WaveTable_test.py append");
  BOOST_CHECK_EQUAL(retval , 0); 
  

}

BOOST_AUTO_TEST_SUITE_END(); 

File: unittest/WaveTable_test.py
#!/usr/bin/python
"""
This is the python file to go along with the WaveTable_test
unit test.

The first argument is the name of the test to run

"""

import tables
import sys

def typecheck(t, name, type):
    assert isinstance(t.coldescrs[name], type)
        
    

def create_test():
    t = tables.openFile('WaveTable_create.h5')
    g = t.root.testGroup
    t = t.root.testGroup.eeg1

    
    assert isinstance(t.coldescrs['src'],
                       tables.description.UInt8Col)
    
    assert isinstance(t.coldescrs['time'],
                       tables.description.UInt64Col)

    typecheck(t, "samprate", tables.description.UInt16Col);
    typecheck(t, "selchan", tables.description.UInt16Col);
    typecheck(t, "filterid", tables.description.UInt16Col);
    assert t.coldescrs["wave"].shape == (128,)
    
    assert t.attrs.src == 17
    
def append_test():
    
    f = tables.openFile('WaveTable_append.h5')
    g = f.root.testGroup
    table = g.eeg2

    N = 1000

    for i in xrange(N):
        assert table[i]['src'] == 23 
        
        assert table[i]['time'] == i * 0x0123456789AB
        assert table[i]['samprate'] == 0x789A
        assert table[i]['selchan'] == 0x1234
        assert table[i]['filterid'] == 0xAABB
        for j in range(128):
            assert table[i]['wave'][j] == i * 0xABCD + j*0x17
        

                
    
    
if __name__ == "__main__":
    if sys.argv[1] == "create":
        create_test()
    elif sys.argv[1] == "append":
        append_test()
        
        
